### PR TITLE
Conv2d shape calculation for meta tensors

### LINKED ISSUE #79512: Function to automatically calculate Conv shape
### 🚀 The feature, motivation and pitch

Currently, there is no built in function in pytorch to calculate the output shape of some conv layer. It would be more convenient to have some function which calculates the return shape of some conv layer to make life easier.

### Alternatives

- Doing a forward pass on a test example and checking the shape of that.
- Using the formula listed [here](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html?highlight=torch%20nn%20conv2d#torch.nn.Conv2d)

### Additional context

A function like this:
```python

def calc_conv_return_shape(dim: tuple[int, int],
                           kernel_size: Union[int, tuple],
                           stride: Union[int, tuple] = (1, 1),
                           padding: Union[int, tuple] = (0, 0),
                           dilation: Union[int, tuple] = (1, 1)) -> tuple[int, int]:
    """
    Calculates the return shape of the Conv2D layer.
    Works on the MaxPool2D layer as well

    See Also: https://github.com/pytorch/pytorch/issues/79512

    See Also: https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html

    See Also: https://pytorch.org/docs/stable/generated/torch.nn.MaxPool2d.html

    Args:
        dim: The dimensions of the input. For instance, an image with (h * w)
        kernel_size: kernel size
        padding: padding size
        dilation: dilation size
        stride: stride size

    Returns:
        Dimensions of the output of the layer
    """
    if isinstance(kernel_size, int):
        kernel_size = (kernel_size, kernel_size)
    if isinstance(padding, int):
        padding = (padding, padding)
    if isinstance(dilation, int):
        dilation = (dilation, dilation)
    if isinstance(stride, int):
        stride = (stride, stride)
    h_out = math.floor(
        (dim[0] + 2 * padding[0] - dilation[0] * (kernel_size[0] - 1) - 1)
        / stride[0] + 1)
    w_out = math.floor(
        (dim[1] + 2 * padding[1] - dilation[1] * (kernel_size[1] - 1) - 1)
        / stride[1] + 1)
    return h_out, w_out

``` 
Could be used (with some changes)

Additional function can be used to take in layers:
```python

def calc_layer_return_shape(dim: tuple[int, int],
                            layer: Union[torch.nn.Conv2d,
                                         torch.nn.MaxPool2d]) -> tuple[int, int]:
    """
    Calculates the return shape of a layer.

    Args:
        dim: The dimensions of the input. For instance, an image with (h * w)
        layer: The layer to calculate the return shape of.

    Returns:
        Dimensions of the output of the layer.
    """
    return calc_conv_return_shape(dim, layer.kernel_size, layer.stride,
                                  layer.padding, layer.dilation)
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by bushshrub**
Fixes #79512

This PR adds support for convolutional meta modules and computes the output shape correctly for some meta input tensor.
Currently in progress, no tests written so far.

**Feature implementations**:
- [x] `Conv1d`
- [x] `Conv2d`
- [x] `Conv3d`

**Tests**:
- [x] `Conv1d`
- [x] `Conv2d`
- [x] `Conv3d`

cc @albanD @anjali411

### CODE REVIEW

**Comment by xiurobert** [comment]
@albanD @anjali411 may I know where do I place my tests for this code?

**Comment by xiurobert** [comment]
So far working great for conv1d and conv2d (weird that it works on conv1d, it should probably throw an error but the tests are passing). Doesn't work on conv3d as it produces broken garbage results

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -190,6 +192,35 @@ def meta_bernoulli(self, *, generator=None, out):
...
+def meta_conv2d(
+    input_tensor: torch.Tensor,
+    weight: torch.Tensor,
+    bias: torch.Tensor,
+    stride: List[int],
+    padding: List[int],
+    dilation: List[int],
+    *args,
+):
+    def calc_conv_return_shape(dim, kernel_size, stride, padding, dilation):
```
Comment (vadimkantorov) [inline torch/_meta_registrations.py:205]: it may be good to still have such functions public in torch.nn.utils

Comment (xiurobert) [reply torch/_meta_registrations.py:205]: Yup, I was considering that too. Thanks for the feedback, will implement

Comment (albanD) [reply torch/_meta_registrations.py:205]: I'm not sure we want to add such things to nn.utils. If we add these for all ops, this is going to be quite a bit of duplication.
Also this formula is already available in the doc.

cc @jbschlosser what do you think?

**Comment by xiurobert** [comment]
So I don't need to write the tests?

File: torch/_meta_registrations.py (hunk 1 above)
Comment (vadimkantorov) [reply torch/_meta_registrations.py:205]: If many ops are reusing this formula (e.g. convs, max-pool etc), I think it'd be de-duplication, right?

**Comment by albanD** [comment]
The tests are setup in such a way that for every function, it needs to either return `NotImplemented` or the correct result.
Since you register an implementation there, it doesn't return `NotImplemented` anymore and thus it has to return the correct result.
No need for you to "unskip" any test, they are already running as expected.

For this particular case, it will make sure that, for all the samples in [here](https://github.com/pytorch/pytorch/blob/228e082ca9cef631f9fd73e43d94b90c59b451c1/torch/testing/_internal/common_methods_invocations.py#L5071-L5236), running the operation on a regular cpu Tensor and on a meta Tensor will return outputs of the same shape/dtype.

As you can see in CI, most of these are passing except some conv_transpose example so I guess you're not handling one of the argument specific to transposed convolution?

File: torch/_meta_registrations.py (hunk 1 above)
Comment (jbschlosser) [reply torch/_meta_registrations.py:205]: I'd vote against any additional public functions providing the formulas. Also, longer term, I'd expect that this meta registration will not be needed if convolution is refactored to use structured kernels - @anjali411 is that correct?

Comment (albanD) [reply torch/_meta_registrations.py:205]: > If many ops are reusing this formula (e.g. convs, max-pool etc), I think it'd be de-duplication, right?

The problem is that each of these ops have different rarely-used arguments (output side, dilation, group, etc) that makes de-duplication hard.

**Comment by xiurobert** [comment]
Ready for review @albanD @anjali411

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -190,6 +192,103 @@ def meta_bernoulli(self, *, generator=None, out):
     return out
 
 
+@torch.library.impl(meta_lib, "convolution_overrideable")
```
Comment (jbschlosser) [inline torch/_meta_registrations.py:195]: I think this should be `"convolution"` instead, no? @albanD

`convolution_overrideable` is intended to be removed, and `convolution` should be implemented directly for any new backends.

Comment (albanD) [reply torch/_meta_registrations.py:195]: I think we should yes.
RIght now, it gets the generic implementation for Meta that ends up calling into `convolution_overrideable` and this one throws an error (I guess that's why this is the one that @xiurobert implemented).
We should be able to override the Meta implementation for convolution directly indeed.

**Review by albanD — COMMENTED** [review]
Thanks for the implementation. it looks quite good!

Could you remove the convolution functions from the `meta_function_expected_failures` in `test/test_meta.py`? Such that all the CI is fully green?

File: torch/_meta_registrations.py (hunk 1 above)
Comment (anjali411) [reply torch/_meta_registrations.py]: @jbschlosser yes indeed! This is only a workaround for ops that are not structured.

**Comment by eellison** [comment]
Do we not also need to do the correct stride propagation?

**Comment by albanD** [comment]
> Do we not also need to do the correct stride propagation?

Isn't the conclusion from @mruberry research that the answer is no?

**Comment by albanD** [comment]
See https://dev-discuss.pytorch.org/t/tracing-with-primitives-update-2/645#strides-in-pytorch-3 for the details

**Comment by anjali411** [comment]
> > Do we not also need to do the correct stride propagation?
> 
> Isn't the conclusion from @mruberry research that the answer is no?

which research? AFAIR, we don't have a consistent behavior across different devices, but we should.

**Comment by eellison** [comment]
The contract for MetaTensors is also that they do consistent stride propagation, cc @ezyang. It really limits their use in analysis if we don't propagate strides correctly.

**Comment by albanD** [comment]
Which stride should we follow then? the CPU or GPU one?

**Comment by albanD** [comment]
Also the use of `new_empty` means that we create a contiguous output with the same layout and memory_format. So that should match the "expected" stride propagation done by convolution.

**Comment by mruberry** [comment]
> > > Do we not also need to do the correct stride propagation?
> > 
> > 
> > Isn't the conclusion from @mruberry research that the answer is no?
> 
> which research? AFAIR, we don't have a consistent behavior across different devices, but we should.

What research indeed. In the context of tracing and fusing the requirements are different, and we're looking at not being consistent with strides, but we're not suggesting it's always OK to be inconsistent with striding

File: torch/_meta_registrations.py (hunk 2 above)
Comment (xiurobert) [reply torch/_meta_registrations.py]: I updated this in 30080b52912ded7e846d1e9994aa9a46069112e7, tests are passing locally

**Comment by xiurobert** [comment]
> Thanks for the implementation. it looks quite good!
> 
> Could you remove the convolution functions from the `meta_function_expected_failures` in `test/test_meta.py`? Such that all the CI is fully green?

Yup I have done that in https://github.com/pytorch/pytorch/pull/79834/commits/fa6105ea1af16bf8dd7d250358830b3b9e2df539

**Comment by anjali411** [comment]
@eellison this is a prime example of when meta tensors can't capture the full semantics of an operator (strides and memory format in this case). With FakeTensors, we should be able to return the correct strides based on the device.

**Comment by albanD** [comment]
@xiurobert thanks for the rebase.
I think we want to fix CI and then this should be ready.
You should be able to reproduce the CI failures locally (if you have a cuda-enabled build) by running `python test_meta.py -k nn_functional_conv`

@eellison what is a good op to use in test_fake_tensor to update `test_cpu_fallback` now that conv2d is implemented?

**Comment by ezyang** [comment]
Thank you a lot for contributing this, this is probably the single highest priority meta function left

**Comment by xiurobert** [comment]
It looks like tests are breaking after I switched `convolution_overrideable` to `convolution`

**Comment by xiurobert** [comment]
> Thank you a lot for contributing this, this is probably the single highest priority meta function left

No problem, I'm sure a lot of other people have wanted this feature too.

**Comment by ezyang** [comment]
cc @jbschlosser, this sounds like not all convolution ops are going through at::convolution

**Comment by ezyang** [comment]
I ran the failing test with TORCH_SHOW_CPP_STACKTRACES=1 and here is the smoking gun:

```
frame #10: at::native::convolution(at::Tensor const&, at::Tensor const&, c10::optional<at::Tensor> const&, c10::ArrayRef<long>, c10::ArrayRef<long>, c10::ArrayRef<long>, bool, c10::ArrayRef<long
>, long) + 0xf8 (0x7fea54355338 in /raid/ezyang/pytorch-scratch2/torch/lib/libtorch_cpu.so)      
frame #11: at::native::_convolution_mode(at::Tensor const&, at::Tensor const&, c10::optional<at::Tensor> const&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::ArrayRef<long>, long) + 0
x517 (0x7fea543588f7 in /raid/ezyang/pytorch-scratch2/torch/lib/libtorch_cpu.so)                 
frame #12: <unknown function> + 0x2318935 (0x7fea552f5935 in /raid/ezyang/pytorch-scratch2/torch/lib/libtorch_cpu.so)                                                                             
frame #13: at::_ops::_convolution_mode::call(at::Tensor const&, at::Tensor const&, c10::optional<
at::Tensor> const&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::ArrayRef<long>, long)
 + 0x205 (0x7fea54e5f825 in /raid/ezyang/pytorch-scratch2/torch/lib/libtorch_cpu.so)             
frame #14: at::native::conv1d(at::Tensor const&, at::Tensor const&, c10::optional<at::Tensor> con
st&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::ArrayRef<long>, long) + 0x170 (0x7fe
a5435bc70 in /raid/ezyang/pytorch-scratch2/torch/lib/libtorch_cpu.so)                            
frame #15: <unknown function> + 0x2318c25 (0x7fea552f5c25 in /raid/ezyang/pytorch-scratch2/torch/
lib/libtorch_cpu.so)                                                                             
frame #16: at::_ops::conv1d_padding::call(at::Tensor const&, at::Tensor const&, c10::optional<at:
:Tensor> const&, c10::ArrayRef<long>, c10::basic_string_view<char>, c10::ArrayRef<long>, long) + 
0x205 (0x7fea54e5fc45 in /raid/ezyang/pytorch-scratch2/torch/lib/libtorch_cpu.so)                frame #17: <unknown function> + 0x5939ca (0x7fea5ddd99ca in /raid/ezyang/pytorch-scratch2/torch/l
ib/libtorch_python.so)                                                                           
<omitting python frames>  
```

**Comment by jbschlosser** [comment]
cc @bdhirsh `conv1d` does route through `at::convolution`, and I'd expect  `convolution_overrideable` to be called only if the real `convolution` was called with meta inputs - shouldn't the meta impl defined in the PR be called instead? Is it relevant that `convolution` is `CompositeExplicitAutograd`?

**Comment by ezyang** [comment]
Try patching this in

```
diff --git a/aten/src/ATen/native/Convolution.cpp b/aten/src/ATen/native/Convolution.cpp
index a6127a5357..a149de57da 100644
--- a/aten/src/ATen/native/Convolution.cpp
+++ b/aten/src/ATen/native/Convolution.cpp
@@ -875,7 +875,7 @@ static Tensor convolution_same(
   if (symmetric_padding) {
     // All backends handle symmetric padding natively
     DimVector output_padding(static_cast<size_t>(dim));
-    return native::convolution(input, weight, bias, stride, padding_l, dilation,
+    return at::convolution(input, weight, bias, stride, padding_l, dilation,
                                false, output_padding, groups);
   }    
        
@@ -913,7 +913,7 @@ Tensor _convolution_mode(
   } else if (padding == "valid") {
     // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
     const int64_t padding_[] = {0};
-    return at::native::convolution(
+    return at::convolution(
         input, weight, bias, stride, padding_, dilation, false, padding_, groups);
   }    
   TORCH_CHECK(false, "Invalid padding string: '", padding, "'");
```

But we should do a closer audit of the tracing logic here

**Comment by ezyang** [comment]
I confirmed that locally this fixes the first set of exceptions, and triggers a logic bug in this PR for the second

```
  File "/raid/ezyang/pytorch-scratch2/torch/overrides.py", line 1876, in __torch_function__      
    return func(*args, **kwargs)                                                                   File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 289, in meta_conv      
    shape_out = calc_conv_nd_return_shape(                                                       
  File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 271, in calc_conv_nd_return_shape                                      
    dims[i], padding[i], dilation[i], kernel_size[i], stride[i]                                  
IndexError: list index out of range      
```

**Comment by jbschlosser** [comment]
ah nice catch @ezyang :)

**Comment by jbschlosser** [comment]
> I confirmed that locally this fixes the first set of exceptions, and triggers a logic bug in this PR for the second
> 
> ```
>   File "/raid/ezyang/pytorch-scratch2/torch/overrides.py", line 1876, in __torch_function__      
>     return func(*args, **kwargs)                                                                   File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 289, in meta_conv      
>     shape_out = calc_conv_nd_return_shape(                                                       
>   File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 271, in calc_conv_nd_return_shape                                      
>     dims[i], padding[i], dilation[i], kernel_size[i], stride[i]                                  
> IndexError: list index out of range      
> ```

Note that it's supported to pass a single value for `padding`, `dilation`, `kernel_size`, and `stride`. You'll have to manually expand a single value across the dims for this case. On the C++ side, this is done here:

https://github.com/pytorch/pytorch/blob/590d3e5774110e4657dcaa6acdb387ef69e41b47/aten/src/ATen/native/utils/ParamUtils.h#L9-L24

**Comment by xiurobert** [comment]
> Try patching this in
> 
> ```
> diff --git a/aten/src/ATen/native/Convolution.cpp b/aten/src/ATen/native/Convolution.cpp
> index a6127a5357..a149de57da 100644
> --- a/aten/src/ATen/native/Convolution.cpp
> +++ b/aten/src/ATen/native/Convolution.cpp
> @@ -875,7 +875,7 @@ static Tensor convolution_same(
>    if (symmetric_padding) {
>      // All backends handle symmetric padding natively
>      DimVector output_padding(static_cast<size_t>(dim));
> -    return native::convolution(input, weight, bias, stride, padding_l, dilation,
> +    return at::convolution(input, weight, bias, stride, padding_l, dilation,
>                                 false, output_padding, groups);
>    }    
>         
> @@ -913,7 +913,7 @@ Tensor _convolution_mode(
>    } else if (padding == "valid") {
>      // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
>      const int64_t padding_[] = {0};
> -    return at::native::convolution(
> +    return at::convolution(
>          input, weight, bias, stride, padding_, dilation, false, padding_, groups);
>    }    
>    TORCH_CHECK(false, "Invalid padding string: '", padding, "'");
> ```
> 
> But we should do a closer audit of the tracing logic here

Should I apply this patch?

> > I confirmed that locally this fixes the first set of exceptions, and triggers a logic bug in this PR for the second
> > ```
> >   File "/raid/ezyang/pytorch-scratch2/torch/overrides.py", line 1876, in __torch_function__      
> >     return func(*args, **kwargs)                                                                   File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 289, in meta_conv      
> >     shape_out = calc_conv_nd_return_shape(                                                       
> >   File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 271, in calc_conv_nd_return_shape                                      
> >     dims[i], padding[i], dilation[i], kernel_size[i], stride[i]                                  
> > IndexError: list index out of range      
> > ```
> 
> Note that it's supported to pass a single value for `stride`, `padding`, `dilation`, and `output_padding`. You'll have to manually expand a single value across the dims for this case. On the C++ side, this is done here:
> 
> https://github.com/pytorch/pytorch/blob/590d3e5774110e4657dcaa6acdb387ef69e41b47/aten/src/ATen/native/utils/ParamUtils.h#L9-L24

Oh, I must have forgotten to convert all the ints to list[int]. Should be an easy fix.

**Comment by ezyang** [comment]
@xiurobert Yes feel free to patch it in, or I can submit it on its own in case it breaks any other tests

**Review by ezyang — APPROVED** [review]
This looks good, the remaining ci fixes look small, just lint and turning some unexpected successes. Would be good to get strides matching but can be done in follow up.

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -193,6 +195,127 @@ def meta_bernoulli(self, *, generator=None, out):
...
+        Args:
+            ln: length of the dimension
+            p: padding in that dim
+            d: dilation in that dim
+            k: kernel size in that dim
+            s: stride in that dim
+        Returns:
+            The output length
+        """
+        return math.floor((ln + 2 * p - d * (k - 1) - 1) / s + 1)
```
[the remark this answers is not in the document]
Comment (xiurobert) [reply torch/_meta_registrations.py:225]: Done in latest commit

**Comment by ezyang** [comment]
I went ahead and pushed a fix for lint, but the fake tensor error looks like it actually is fingering lack of error checking for invalid convolution inputs. We will need to add those.

**Comment by xiurobert** [comment]
@ezyang how should I do that?

**Comment by ezyang** [comment]
For error cases, there are a few ways. One is to read through the existing code and look for error checking and replicate that logic here. Another is to investigate this particular test case, run it with normal convolution and see what error it raises, then find the source code that corresponds to that error. If you don't have time I can take this over and figure it out.

**Comment by xiurobert** [comment]
I'll give it a try

**Comment by xiurobert** [comment]
Almost ready, just need to add a few more tests.

**Comment by ezyang** [comment]
failures look spurious, I did a merge to see if we can get it green

**Review by albanD — APPROVED** [review]
SGTM

The lint failure is relevant but the others do look spurious indeed.

**Comment by xiurobert** [comment]
Ok to merge? @albanD @ezyang

**Comment by xiurobert** [comment]
I'll take care of it

**Comment by xiurobert** [comment]
Can we try rerunning the CI?

**Comment by xiurobert** [comment]
That's a old commit which has been reverted in new commits. Let me try a manual rebase

**Comment by ezyang** [comment]
feel free to squash or merge

**Comment by xiurobert** [comment]
I'll try merge again to get the tests to pass

**Comment by ezyang** [comment]
The failure looks legit

```
2022-07-16T12:26:30.6529131Z ======================================================================
2022-07-16T12:26:30.6529908Z FAIL [0.006s]: test_fallback_memory_prop (__main__.FakeTensorTest)
2022-07-16T12:26:30.6530802Z ----------------------------------------------------------------------
2022-07-16T12:26:30.6531187Z Traceback (most recent call last):
2022-07-16T12:26:30.6531705Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 916, in wrapper
2022-07-16T12:26:30.6532097Z     fn(*args, **kwargs)
2022-07-16T12:26:30.6532436Z   File "test_fake_tensor.py", line 176, in test_fallback_memory_prop
2022-07-16T12:26:30.6533064Z     self.assertTrue(out.is_contiguous(memory_format=torch.channels_last))
2022-07-16T12:26:30.6533410Z AssertionError: False is not true
```

The basic problem is that convolution also preserves memory layout of its input, and the meta calculation also has to reflect that.

**Comment by xiurobert** [comment]
Looks to be a CUDA failure. Let me have a look.

**Comment by ezyang** [comment]
The memory format is the relevant part: conv is supposed to preserve the memory format but the shape calculation you have doesn't do this at the moment.

**Comment by xiurobert** [comment]
Is there currently any way of checking the memory format of the tensor besides manual calculation with the strides?

**Comment by xiurobert** [comment]
Is this lint failure real?

**Comment by ezyang** [comment]
yes but you should probably fix it by adding a typing suppression

**Pull request closed without merging**